lib/puppet/transaction/report.rb

```diff
@@ -308,14 +308,23 @@ def initialize_from_hash(data)
     end
   end
 
-  def skip_or_to_data_hash(rs)
-    if rs.nil?
-      nil
-    elsif !rs.out_of_sync? && Puppet[:exclude_unchanged_resources]
-      {}
-    else
-      rs.to_data_hash
+  def keep_resource_status?(statuses, rs)
+    statuses.each do |status|
+      if rs.send(status) == true
+        return true
+      end
     end
+      return false
+  end
```

Comment:
There's kind of an O^2 thing (scanning the list of `statuses` for every resource status). But the list of `statuses` should be small... maybe 4 at the most, so it's more like 4 * number of resource statuses. I think that's ok.

```suggestion
  statuses.any? do |status|
    rs.send(status)
  end
```

Reply:
I'm not sure if this code snippet is necessary because we only execute this path if `statuses_to_omit` is not empty. I think it might be a little confusing here because I think I am conflating `status` with the status attribute of `STATE`, so I will try to clean that up a bit to be less confusing; I'm pretty sure we don't need this code snippet though...

Reply:
oh wait, if `Puppet[:resource_statuses_to_omit] = "skipped, out_of_sync"`, then we'll want to keep the status if none of the statuses are `true`. So I think like:

```ruby
def keep_resource_status?(statuses, rs)
  statuses.none? do |status|
    rs.send(status)
  end
end
```
If `statuses` is empty, then it's `true`:

```
irb(main):006:0> [].none? {}
=> true
```

I assume all of the status related reader methods, e.g. `Status#failed`, return a boolean so we don't need to double negate `!!.rs.send(status)`?